Restructure torchao quantization examples

jerryzh168 opened the pull request:
Summary:
Mainly structured the examples by hardwares and then listed the recommended quantization methods for each hardware H100 GPU, A100 GPU and CPU

Also added example for push_to_hub

Test Plan:
not required

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Restructure torchao quantization examples

Summary:
Mainly structured the examples by hardwares and then listed
the recommended quantization methods for each hardware H100 GPU, A100 GPU and CPU

Also added example for push_to_hub

Test Plan:
not required

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 commented:
cc @MekkCyber @SunMarc please take a look

one thing is I'm not sure if people use float8 on CPU, haven't heard of this before actually

jerryzh168 committed:
update

jerryzh168 committed:
drop float8 cpu

jerryzh168 commented:
@MekkCyber I dropped float8 cpu example since that's not commonly used and doesn't yield speedup according to benchmarks done by @jainapurva

MekkCyber commented on docs/source/en/quantization/torchao.md:
````diff
@@ -63,7 +64,7 @@ pip install --upgrade torchao transformers
 <hfoption id="PyTorch Index">
 Stable Release from the PyTorch index
 ```bash
-pip install torchao --extra-index-url https://download.pytorch.org/whl/cu124 # options are cpu/cu118/cu124/cu126
+pip install torchao --extra-index-url https://download.pytorch.org/whl/cu126 # options are cpu/cu118/cu126/cu128
````
In the PyPi installation, I specified that running `pip install torchao` will default to 12.4 according to the doc in comment, but I assume this is no longer the case, can you please update that too (line 60)

MekkCyber commented on docs/source/en/quantization/torchao.md:
````diff
@@ -346,11 +388,24 @@ torchao implements [torch.Tensor subclasses](https://pytorch.org/docs/stable/not
+</hfoption>
+<hfoption id="push to huggingface hub">
+```py
+# don't serialize model with Safetensors
+USER_ID = # YOUR HUGGINGFACE USER ID
+REPO_ID = "llama3-8b-int4wo-128"
+quantized_model.push_to_hub(f"{USER_ID}/llama3-8b-int4wo-128", safe_serialization=False)
+tokenizer.push_to_hub(f"{USER_ID}/llama3-8b-int4wo-128")
+```
+</hfoption>
+
````
This is causing the doc pr to fail, can you replace it with 
```suggestion
# don't serialize model with Safetensors
USER_ID = "your_huggingface_user_id"
REPO_ID = "llama3-8b-int4wo-128"
quantized_model.push_to_hub(f"{USER_ID}/llama3-8b-int4wo-128", safe_serialization=False)
tokenizer.push_to_hub(f"{USER_ID}/llama3-8b-int4wo-128")
```
</hfoption>

```

MekkCyber commented on docs/source/en/quantization/torchao.md:
````diff
@@ -164,45 +174,42 @@ print(tokenizer.decode(output[0], skip_special_tokens=True))
 ```
 </hfoption>
-<hfoption id="int8-dynamic-quantization cuda">
+<hfoption id="int8-weight-only cuda">
````
no need to specify cuda since there are A100 examples
```suggestion
<hfoption id="int8-weight-only">
```

MekkCyber commented on docs/source/en/quantization/torchao.md:
```diff
@@ -222,14 +229,18 @@ output = quantized_model.generate(**input_ids, max_new_tokens=10, cache_implemen
+
+### CPU
+<hfoptions id="examples-CPU">
+<hfoption id="int8-weight-only cpu">
```
```suggestion
<hfoption id="int8-weight-only">
```

MekkCyber commented on docs/source/en/quantization/torchao.md:
````diff
@@ -249,14 +260,45 @@ output = quantized_model.generate(**input_ids, max_new_tokens=10, cache_implemen
 ```
 </hfoption>
-<hfoption id="float8-weight-only cuda">
+<hfoption id="int4-weight-only cpu">
````
```suggestion
<hfoption id="int4-weight-only">
```

MekkCyber commented on docs/source/en/quantization/torchao.md:
````diff
@@ -249,14 +260,45 @@ output = quantized_model.generate(**input_ids, max_new_tokens=10, cache_implemen
+print(tokenizer.decode(output[0], skip_special_tokens=True))
+```
+</hfoption>
+<hfoption id="int8-dynamic-quantization cuda">
+
+```py
+import torch
+from transformers import TorchAoConfig, AutoModelForCausalLM, AutoTokenizer
+from torchao.quantization import Int8DynamicActivationInt8WeightConfig
+
+quant_config = Int8DynamicActivationInt8WeightConfig()
````
this is mistakenly left with the cpu examples, to determine if it's better to add it with H100 or A100 examples, I think A100 would be a better fit

jerryzh168 replied:
yeah A100 would be better, I also want to add a gemlite examples for 4 bit for H100 as well

jerryzh168 replied:
also planning to simplify this to just float8, int8, int4 as well

MekkCyber commented on docs/source/en/quantization/torchao.md:
````diff
@@ -274,17 +316,16 @@ input_ids = tokenizer(input_text, return_tensors="pt").to("cuda")
 ```
 </hfoption>
-<hfoption id="float8-weight-only cpu">
+<hfoption id="int8-dynamic-quantization cpu">
````
```suggestion
<hfoption id="int8-dynamic-quantization">
```

MekkCyber reviewed (approved):
Huge thanks @jerryzh168 for adding the new examples 🤗  —especially the GemLite one! It’ll be super helpful for a lot of people. 

Concerning FP8 for cpu no problem for dropping it. 

it looks great I just left some few nits

SunMarc reviewed (approved):
Thanks a lot for updating the docs ! Looks cleaner this way. Please resolve @MekkCyber comments and we should be good to merge

jerryzh168 committed:
address comments and simplify

jerryzh168 committed:
small update

jerryzh168 committed:
link update

jerryzh168 commented:
Thanks for the review @MekkCyber and @SunMarc, I addressed the comments and also merged dynamic and weight only quant examples to make it simper, now people mainly see:
```
hardware:
1 example for 8 bit (int8/float8)
1 example for 4 bit (int4)
```
and there will be more details inline, either dynamic/weight only or batch size 1/N

jerryzh168 committed:
minor update
